src/edu/csus/ecs/pc2/core/execute/Executable.java

```diff
@@ -1811,7 +1846,25 @@ protected boolean executeProgram(int dataSetNumber) {
 
             }
 
-            log.log(Log.DEBUG, "before substitution: " + cmdline);
+            //determine whether the Problem is configured to use a sandbox 
+            try {
+                usingSandbox = isUsingSandbox();
+            } catch (Exception e) {
+                //an exception means there is something wrong about sandbox usage (e.g. not allowed on this platform)
+                log.severe("Exception during sandbox usage check: " + e.getMessage());
+                stderrlog.close();
+                stdoutlog.close();
+                executionData.setExecuteSucess(false);
+                executionData.setExecutionException(e);
+                return false;
+            }
+            
+            if (usingSandbox) {
```

Comment:
Note the TODO comments in method `isUsingSandbox()`, which is called (above) to set the value of the variable `isUsingSandbox`:
```java
    /**
     * Returns true if the current Problem is configured to use a sandbox, sandbox usage is supported on the current platform, 
     * and the current client is not a Team; false otherwise.
     * If the returned value is True, the sandbox for the problem will have been copied into the Execute directory.
     * 
     *TODO: this is a query method with a side-effect (copying the sandbox into the execute directory); that's not good style...
     *TODO:  Need to refactor this into two steps:  a query (without side effects) asking whether the problem is supposed to use
     *TODO:    a sandbox, and then if that returns True, a call to a separate method which installs the sandbox.  This is tricky
     *TODO:    because we want to get the proper exception back to the caller...
     * 
     * @return true if the Problem is to use a sandbox; false if not.  If True is returned, the sandbox will have been copied into 
     *              the Execute directory.
     * 
     * @throws Exception if there is a sandbox configuration issue, such as 
     *          we're running on an OS platform where sandbox isn't supported, or
     *          the specified sandbox can't be loaded into the execute directory.
     */
```
I would _really_ like to see method `isUsingSandbox()` corrected so that it has no side-effects (query accessors shouldn't have side effects).  Instead, the code inside `isUsingSandbox()` which actually copies the sandbox into the execute directory should be refactored into its own method, which should then be called at this point.  It does probably mean adding an additional `try/catch` around the code here, but I think it would be worth it to avoid introducing "bad things" into the code.  (And yes, I'm totally aware that _I'm the one who wrote it that way in the first place :)_ ... but at least I put a big fat comment pointing out that it should be fixed. ) I'd sure like to see it fixed in the first version of the code which goes into production...

Reply:
Fixed in  [this commit](https://github.com/pc2ccs/pc2v9/pull/732/commits/376601787954cfe6acce71cc171ee64b1e9241b0)